rules_gapic/go/go_gapic.bzl

```diff
@@ -0,0 +1,174 @@
+    )
+
+_go_gapic_postprocessed_srcjar = go_rule(
+    _go_gapic_postprocessed_srcjar_impl,
+    attrs = {
+        "gapic_srcjar": attr.label(mandatory = True, allow_single_file = True),
+    },
+    outputs = {
+        "main": "%{name}.srcjar",
+        "test": "%{name}-test.srcjar",
+        "smoke_test": "%{name}-smoke-test.srcjar",
+    },
+)
+
+def go_gapic_srcjar(name, src, gapic_yaml, service_yaml, visibility = None):
+    raw_srcjar_name = "%s_raw" % name
+
+    gapic_srcjar(
+        name = raw_srcjar_name,
+        src = src,
+        gapic_yaml = gapic_yaml,
+        service_yaml = service_yaml,
+        visibility = visibility,
+        artifact_type = "GAPIC_CODE",
+        language = "go",
+    )
+
+    _go_gapic_postprocessed_srcjar(
+        name = name,
+        gapic_srcjar = ":%s" % raw_srcjar_name,
+        visibility = visibility,
+    )
+
+def go_gapic_library(
+        name,
+        src,
+        gapic_yaml,
+        service_yaml,
+        importpath,
+        deps = [],
+        visibility = None):
+    srcjar_name = "%s_srcjar" % name
+    # TODO: make smoke_tests compilable (remove /internal/ dependency from generated code)
+    # Smoke_test_library_name = "%s_smoke_test" % name
+
+    go_gapic_srcjar(
+        name = srcjar_name,
+        src = src,
+        gapic_yaml = gapic_yaml,
+        service_yaml = service_yaml,
+        visibility = visibility,
+    )
+
+    actual_deps = deps + [
+        "@com_github_googleapis_gax_go//v2:go_default_library",
+        "@org_golang_google_api//option:go_default_library",
+        "@org_golang_google_api//iterator:go_default_library",
+        "@org_golang_google_api//transport:go_default_library",
+        "@org_golang_google_grpc//:go_default_library",
+        "@org_golang_google_grpc//codes:go_default_library",
+        "@org_golang_google_grpc//metadata:go_default_library",
+        "@com_github_golang_protobuf//proto:go_default_library",
+        "@com_github_golang_protobuf//ptypes:go_default_library",
+        "@com_github_golang_protobuf//ptypes/empty:go_default_library",
+        "@com_github_golang_protobuf//ptypes/timestamp:go_default_library",
+        "@org_golang_google_genproto//protobuf/field_mask:go_default_library",
+        "@com_google_googleapis//google/rpc:status_go_proto",
+        "@org_golang_google_grpc//status:go_default_library",
+    ]
+
+    main_file = ":%s.srcjar" % srcjar_name
+    main_dir = "%s_main" % srcjar_name
+
+    unzipped_srcjar(
+        name = main_dir,
+        srcjar = main_file,
+        extension = ".go",
+    )
+
+    go_library(
+        name = name,
+        srcs = [":%s" % main_dir],
+        deps = actual_deps,
+        importpath = importpath,
+    )
+
+    test_file = ":%s-test.srcjar" % srcjar_name
+    test_dir = "%s_test" % srcjar_name
+
+    unzipped_srcjar(
+        name = test_dir,
+        srcjar = test_file,
+        extension = ".go",
+    )
+
+#    smoke_test_file = ":%s-smoke-test.srcjar" % srcjar_name
```

Comment:
We currently do not compile and/or run smoke tests (it is impossible with rules go to compile but not run tests, at the same time we cannot run smoke tests, because they require additional preparaion). This is mostly for "future use". I added comment explaining this.